src/montgomery.rs

```diff
@@ -150,34 +193,256 @@ impl CompressedMontgomeryU {
+/// include the sign bit, there's two points here (if translated from the
+/// Edwards form) which will equate.
+///
+/// # Returns
+///
+/// `1` if the points are equal, and `0` otherwise.
+impl Equal for MontgomeryPoint {
+    fn ct_eq(&self, that: &MontgomeryPoint) -> u8 {
+        slices_equal(self.compress().as_bytes(),
+                     that.compress().as_bytes())
+    }
+}
+
+/// Determine if this `MontgomeryPoint` is valid.
+///
+/// # Note
+///
+/// All points, except for `(X:W) = (0:0)`, are valid, since the projective
+/// model is linear through the origin and is comprised by all `X` in
+/// ℤ/(2²⁵⁵-19).
+///
+/// # Returns
+///
+/// `true` if it is valid, and `false` otherwise.
+impl ValidityCheck for MontgomeryPoint {
+    fn is_valid(&self) -> bool {
+        let zero = FieldElement::zero();
+
+        if (self.U.ct_eq(&zero) & self.W.ct_eq(&zero)) == 1 {
+            return true;
+        }
+        false
+    }
+}
+
+/// Conditionally assign another `MontgomeryPoint` to this point, in constant time.
+///
+/// If `choice == 1`, assign `that` to `self`.  Otherwise, leave `self`
+/// unchanged.
+impl ConditionallyAssignable for MontgomeryPoint {
+    fn conditional_assign(&mut self, that: &MontgomeryPoint, choice: Mask) {
+        self.U.conditional_assign(&that.U, choice);
+        self.W.conditional_assign(&that.W, choice);
+    }
+}
+
+impl MontgomeryPoint {
+    /// Compress this point to only its u-coordinate (note: affine).
+    ///
+    /// # Returns
+    ///
+    /// A `CompressedMontgomeryU`.
+    pub fn compress(&self) -> CompressedMontgomeryU {
+        let u_affine: FieldElement = &self.U * &self.W.invert();
+
+        CompressedMontgomeryU(u_affine.to_bytes())
+    }
+
+    /// Differential addition for single-coordinate Montgomery points.
+    ///
+    /// Montgomery coordinates in projective 𝗣¹ space are odd in that 𝗣¹
+    /// inherits none of the group structure from E_(A,B).  Hence, the mapping
+    /// of the group operation, `⊕`, is undefined for the pair `(x(P), x(Q))`;
+    /// that is, given `x(P)` and `x(Q)`, we cannot derive `x(P ⊕ Q)`.  This is
+    /// due to the fact that, in Montgomery coordinates, `x(P)` determines `P`
+    /// only up to a sign, and thus we cannot differentiate `x(P ⊕ Q)` from
+    /// `x(P ⊖ Q)`.  However, via differential addition, any three of the values
+    /// `{x(P), x(Q), x(P ⊕ Q), x(P ⊖ Q)}` determines the forth, so we can
+    /// define *pseudo-addition* for a singular coordinate.
+    ///
+    /// # Warning
+    ///
+    /// If the `difference` is the identity point, or a two torsion point, the
+    /// results of this method are not correct, but instead result in `(0:0)`
+    /// (an invalid projective point in the Montgomery model).
+    ///
+    /// The doubling case is degenerate, in that using this method to accomplish
+    /// point doubling is less efficient than using `differential_double()`.
```

Comment:
this can't be used for doubling at all, since then the difference would be the identity...

Reply:
ca5b58c2b75